src/Spice86.Core/Emulator/InterruptHandlers/Common/IndexBasedDispatcher/IndexBasedDispatcher.cs

```diff
@@ -0,0 +1,86 @@
+namespace Spice86.Core.Emulator.InterruptHandlers.Common.IndexBasedDispatcher;
+
+using Spice86.Core.Emulator.Errors;
+using Spice86.Core.Emulator.VM;
+using Spice86.Shared.Interfaces;
+
+using System.Linq;
+
+/// <summary>
+/// Base class for most classes having to dispatch operations depending on a numeric value, like interrupts.
+/// </summary>
+public abstract class IndexBasedDispatcher<T> where T: IRunnable {
+    /// <summary>
+    /// Defines all the available runnables. Each one has an index number and code associated with it.
+    /// </summary>
+    private Dictionary<int, T> _dispatchTable = new();
+
+    /// <summary>
+    /// The logger service implementation.
+    /// </summary>
+    protected readonly ILoggerService _loggerService;
+
+    /// <summary>
+    /// The emulator machine.
+    /// </summary>
+    protected readonly Machine _machine;
+
+    /// <summary>
+    /// Initializes a new instance of an <see cref="IndexBasedDispatcher"/>
+    /// </summary>
+    /// <param name="machine">The emulator machine instance</param>
+    /// <param name="loggerService">The logging service to be used (eg. for recording warnings or errors).</param>
+    public IndexBasedDispatcher(Machine machine, ILoggerService loggerService) {
+        _machine = machine;
+        _loggerService = loggerService;
+    }
+
+    /// <summary>
+    /// Extract the list of runnable items registered
+    /// </summary>
+    /// <returns>A list of all the runnables registered</returns>
+    public IList<T> GetAllRunnables() {
```

Comment:
Maybe we could keep it as IEnumerable.

LINQ's Deferred execution means that the evaluation of an expression is delayed until its realized value is actually required. It greatly improves performance by avoiding unnecessary execution.

ToList executes the query right now and makes LINQ code less performant. As a general rule, ToList, ToArray and friends should be avoided.

Reply:
Good idea, I am a bit lost with LINQ, need more practice.